### PR TITLE
ARTEMIS-3808 support start/stop embedded web server via mngmnt

### PR DESCRIPTION
**Pull request by jbertram**
It would be useful to be able to cycle the embedded web server if, for
example, one needed to renew the SSL certificates. To support this
functionality I made a handful of changes, e.g.:

 - Refactoring `WebServerComponent` so that all the necessary
   configuration would happen in the `start()` method.
 - Refactoring `WebServerComponentTest` to re-use code.

### CODE REVIEW

File: artemis-core-client/src/main/java/org/apache/activemq/artemis/api/core/management/ActiveMQServerControl.java [hunk 1]
```diff
@@ -1971,5 +1971,11 @@ void replay(@Parameter(name = "startScanDate", desc = "Start date where we will
                @Parameter(name = "address", desc = "Name of the address to replay") String address,
                @Parameter(name = "target", desc = "Where the replay data should be sent") String target,
                @Parameter(name = "filter", desc = "Filter to apply on message selection. Null means everything matching the address") String filter) throws Exception;
+
+   @Operation(desc = "stop the embedded web server", impact = MBeanOperationInfo.ACTION)
+   void stopEmbeddedWebServer() throws Exception;
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/api/core/management/ActiveMQServerControl.java:1976]: With the current change, if you were using the web console to stop the embedded web server then you'd have to use something _other_ than the web console to start it again (e.g. JConsole, Jolokia, etc.). However, it does make sense to have a method to both stop & start it. The user experience would certainly be better with that method in most circumstances. I'll add that.

[the remark this answers is not in the document]
Comment (clebertsuconic) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/api/core/management/ActiveMQServerControl.java:1976]: @jbertram Yeah.. It would be nice (and I think simple) to have it.

[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/api/core/management/ActiveMQServerControl.java:1976]: @clebertsuconic done

**Comment by clebertsuconic** [comment]
@jbertram I did some simple testing.. and the console would not work after a restart:

```

022-05-02 16:52:16,600 INFO  [org.apache.activemq.hawtio.branding.PluginContextListener] Destroyed activemq-branding plugin
2022-05-02 16:52:31,616 INFO  [io.hawt.web.auth.AuthenticationFilter] Failed to invoke action /exec/org.apache.activemq.artemis:broker="0.0.0.0"/restartEmbeddedWebServer() due to:: java.security.PrivilegedActionException: org.eclipse.jetty.io.EofException
	at java.base/java.security.AccessController.doPrivileged(Native Method) [java.base:]
	at java.base/javax.security.auth.Subject.doAs(Subject.java:423) [java.base:]
	at io.hawt.web.auth.AuthenticationFilter.executeAs(AuthenticationFilter.java:104) [hawtio-system-2.14.2.jar:2.14.2]
	at io.hawt.web.auth.AuthenticationFilter.doFilter(AuthenticationFilter.java:72) [hawtio-system-2.14.2.jar:2.14.2]
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)
	at io.hawt.web.filters.HttpHeaderFilter.doFilter(HttpHeaderFilter.java:43) [hawtio-system-2.14.2.jar:2.14.2]
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)
	at io.hawt.web.filters.HttpHeaderFilter.doFilter(HttpHeaderFilter.java:43) [hawtio-system-2.14.2.jar:2.14.2]
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)
	at io.hawt.web.filters.HttpHeaderFilter.doFilter(HttpHeaderFilter.java:43) [hawtio-system-2.14.2.jar:2.14.2]
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)
	at io.hawt.web.filters.HttpHeaderFilter.doFilter(HttpHeaderFilter.java:43) [hawtio-system-2.14.2.jar:2.14.2]
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)
	at io.hawt.web.filters.HttpHeaderFilter.doFilter(HttpHeaderFilter.java:43) [hawtio-system-2.14.2.jar:2.14.2]
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)
	at io.hawt.web.filters.HttpHeaderFilter.doFilter(HttpHeaderFilter.java:43) [hawtio-system-2.14.2.jar:2.14.2]

```


I don't think this is related to the restart method. I think there's something else happening there.

**Comment by jbertram** [comment]
If I run `restartEmbeddedWebServer` from JConsole everything works just fine. However, if I run it from the web console it doesn't work. The web server never restarts. I think the problem is that the thread servicing the request from the web console is controlled by Jetty and since Jetty is shutting down the thread just halts. It's a catch-22. I'll try running it in its own thread and see if that works.

**Comment by jbertram** [comment]
@clebertsuconic looks like that was the culprit. Try now. In my testing it's reliable now from JConsole and the web console as well.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/ActiveMQServerControlImpl.java [hunk 2]
```diff
@@ -4479,5 +4482,39 @@ public void replay(String startScan, String endScan, String address, String targ
 
       server.replay(startScanDate, endScanDate, address, target, filter);
    }
+
+   @Override
+   public void stopEmbeddedWebServer() throws Exception {
+      for (ActiveMQComponent component : server.getExternalComponents()) {
```
Comment (clebertsuconic) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/ActiveMQServerControlImpl.java:4488]: shouldn't this be on its own thread as well? what if you stop from the console?

don't make sense I know, but wouldn't be broken if a user do it?

**Comment by clebertsuconic** [comment]
I will test tomorrow.. (including on stopping from the console)

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/ActiveMQServerControlImpl.java (hunk 2 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/ActiveMQServerControlImpl.java:4488]: Stopping the embedded web server from the web console is not a problem. It's having the thread survive so that it can _start_ it again that's the problem.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java [hunk 3]
```diff
@@ -4290,6 +4292,65 @@ public boolean isStarted() {
...
+         serverControl.stopEmbeddedWebServer();
+         Assert.assertFalse(fake.isStarted());
+         serverControl.startEmbeddedWebServer();
+         Assert.assertTrue(fake.isStarted());
+      } else {
+         Assert.assertTrue(System.currentTimeMillis() > fake.getStartTime());
+         long time = System.currentTimeMillis();
+         Thread.sleep(1);
+         serverControl.restartEmbeddedWebServer();
+         Assert.assertTrue(time < fake.getStartTime());
```
[the remark this answers is not in the document]
Comment (jbertram) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java:4365]: Good point.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java [hunk 4]
```diff
@@ -4290,6 +4292,65 @@ public boolean isStarted() {
...
+
+   private void testStartStopEmbeddedWebServer(boolean manual) throws Exception {
+      class Fake implements ServiceComponent, WebServerComponentMarker {
+         boolean started = false;
+         long startTime = 0;
+
+         @Override
+         public void start() throws Exception {
+            started = true;
+            startTime = System.currentTimeMillis();
```
[the remark this answers is not in the document]
Comment (jbertram) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java:4317]: Done.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java [hunk 5]
```diff
@@ -4290,6 +4292,65 @@ public boolean isStarted() {
...
+         }
+
+         @Override
+         public void stop() throws Exception {
+         }
+
+         @Override
+         public void stop(boolean shutdown) throws Exception {
+            started = false;
+         }
```
[the remark this answers is not in the document]
Comment (jbertram) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java]: Done.

File: artemis-web/src/test/java/org/apache/activemq/cli/test/WebServerComponentTest.java [hunk 6]
```diff
@@ -175,31 +166,31 @@ public void testComponentStopBehavior() throws Exception {
...
-      assertEquals("12345", clientHandler.body.toString());
-      // Wait for the server to close the connection.
-      ch.close();
-      ch.eventLoop().shutdownNow();
+      webServerComponent.stop(true);
+      Assert.assertFalse(webServerComponent.isStarted());
+   }
+
+   @Test
+   public void testComponentStopStartBehavior() throws Exception {
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-web/src/test/java/org/apache/activemq/cli/test/WebServerComponentTest.java:182]: I added docs.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/ActiveMQServerControlImpl.java [hunk 7]
```diff
@@ -4479,5 +4483,45 @@ public void replay(String startScan, String endScan, String address, String targ
...
+   @Override
+   public void restartEmbeddedWebServer() throws Exception {
+      /*
+       * This needs to be run in its own thread managed by the broker because if it is run on a thread managed by Jetty
+       * (e.g. if it is invoked from the web console) then the thread will die before Jetty can be restarted.
+       */
+      server.getThreadPool().execute(() -> {
+         try {
+            stopEmbeddedWebServer();
+            startEmbeddedWebServer();
```
Comment (gemmellr) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/ActiveMQServerControlImpl.java:4506]: Thinking about this more I still think these need coordinated here in some way. If I call restart, and then I immediately call stop from the same thread, the webserver should end up stopped. As this is now, because they are handled in different threads it seems just as likely it will end up [re]started instead.

Regardless of approach taken to addressing that, it feels like maybe it should even also await the operation being completed for some period before returning. It could even throw on failure like start and stop would.

Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/ActiveMQServerControlImpl.java]: Done.

File: artemis-web/src/main/java/org/apache/activemq/artemis/component/WebServerComponent.java [hunk 8]
```diff
@@ -193,87 +203,97 @@ public void configure(ComponentDTO config, String artemisInstance, String artemi
...
+      connector.setHost(uri.getHost());
+      connector.setName("Connector-" + i);
+      return connector;
    }
 
    public void internalStop() throws Exception {
+      ActiveMQWebLogger.LOGGER.stoppingEmbeddedWebServer();
+      System.out.println("Stopping...");
       server.stop();
+      server = null;
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-web/src/main/java/org/apache/activemq/artemis/component/WebServerComponent.java]: Done.

File: artemis-web/src/main/java/org/apache/activemq/artemis/component/WebServerComponent.java [hunk 9]
```diff
@@ -193,87 +203,97 @@ public void configure(ComponentDTO config, String artemisInstance, String artemi
...
+         connector = new ServerConnector(server, connectionFactory);
+      }
+      connector.setPort(uri.getPort());
+      connector.setHost(uri.getHost());
+      connector.setName("Connector-" + i);
+      return connector;
    }
 
    public void internalStop() throws Exception {
+      ActiveMQWebLogger.LOGGER.stoppingEmbeddedWebServer();
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-web/src/main/java/org/apache/activemq/artemis/component/WebServerComponent.java]: Resolved by moving this into `stop(boolean)`.

File: artemis-web/src/main/java/org/apache/activemq/artemis/component/WebServerComponent.java [hunk 10]
```diff
@@ -53,115 +54,122 @@
...
+   public static final String DIR_ALLOWED = "org.eclipse.jetty.servlet.Default.dirAllowed";
 
    private Server server;
    private HandlerList handlers;
-   private WebServerDTO webServerConfig;
+   private WebServerDTO config;
    private final List<String> consoleUrls = new ArrayList<>();
    private final List<String> jolokiaUrls = new ArrayList<>();
-   private List<WebAppContext> webContexts;
+   private List<WebAppContext> webContexts = new ArrayList<>();;
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-web/src/main/java/org/apache/activemq/artemis/component/WebServerComponent.java]: Done.

File: artemis-web/src/main/java/org/apache/activemq/artemis/component/WebServerComponent.java [hunk 11]
```diff
@@ -193,87 +203,97 @@ public void configure(ComponentDTO config, String artemisInstance, String artemi
...
+      }
+      connector.setPort(uri.getPort());
+      connector.setHost(uri.getHost());
+      connector.setName("Connector-" + i);
+      return connector;
    }
 
    public void internalStop() throws Exception {
+      ActiveMQWebLogger.LOGGER.stoppingEmbeddedWebServer();
+      System.out.println("Stopping...");
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-web/src/main/java/org/apache/activemq/artemis/component/WebServerComponent.java]: Removed.

File: artemis-web/src/main/java/org/apache/activemq/artemis/component/WebServerComponent.java [hunk 12]
```diff
@@ -53,115 +54,122 @@
...
-      List<BindingDTO> bindings = webServerConfig.getBindings();
-      connectors = new ServerConnector[bindings.size()];
-      String[] virtualHosts = new String[bindings.size()];
-
-      for (int i = 0; i < bindings.size(); i++) {
-         BindingDTO binding = bindings.get(i);
-         URI uri = new URI(binding.uri);
-         String scheme = uri.getScheme();
-         ServerConnector connector;
+   private RequestLogHandler getLogHandler() {
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-web/src/main/java/org/apache/activemq/artemis/component/WebServerComponent.java]: I think this is mainly due to the way the diff algorithm works, but I rearranged some stuff to eliminate it.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java [hunk 13]
```diff
@@ -4290,6 +4292,80 @@ public boolean isStarted() {
...
+   @Test
+   public void testRestartEmbeddedWebServer() throws Exception {
+      testStartStopEmbeddedWebServer(false);
+   }
+
+   private void testStartStopEmbeddedWebServer(boolean manual) throws Exception {
+      class Fake implements ServiceComponent, WebServerComponentMarker {
+         boolean started = false;
+         long startTime = 0;
+         long stopTime = 0;
```
[the remark this answers is not in the document]
Comment (jbertram) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java]: Resolved.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java [hunk 14]
```diff
@@ -4290,6 +4292,80 @@ public boolean isStarted() {
...
+         Assert.assertTrue(fake.isStarted());
+      } else {
+         // this is technically async so need to do some gymnastics to test effectively
+         long time = System.currentTimeMillis();
+         Thread.sleep(5);
+         Assert.assertTrue(time > fake.getStartTime());
+         Assert.assertTrue(time > fake.getStopTime());
+         serverControl.restartEmbeddedWebServer();
+         Wait.assertTrue(() -> time < fake.getStopTime(), 5000, 100);
+         Wait.assertTrue(() -> serverControl.isEmbeddedWebServerStarted(), 5000, 100);
```
[the remark this answers is not in the document]
Comment (jbertram) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java]: Should be resolved now.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/ActiveMQServerControlImpl.java [hunk 15]
```diff
@@ -4479,5 +4483,45 @@ public void replay(String startScan, String endScan, String address, String targ
...
+         } catch (Exception e) {
+            ActiveMQServerLogger.LOGGER.failedToRestartEmbeddedWebServer(e);
+         }
+      });
+   }
+
+   @Override
+   public boolean isEmbeddedWebServerStarted() throws Exception {
+      return getEmbeddedWebServerComponent().isStarted();
+   }
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/ActiveMQServerControlImpl.java]: Agreed.

File: artemis-web/src/main/java/org/apache/activemq/artemis/component/WebServerComponent.java [hunk 16]
```diff
@@ -363,9 +366,17 @@ public void stop() throws Exception {
    }
 
    @Override
-   public void stop(boolean isShutdown) throws Exception {
+   public synchronized void stop(boolean isShutdown) throws Exception {
       if (isShutdown) {
-         internalStop();
+         ActiveMQWebLogger.LOGGER.stoppingEmbeddedWebServer();
+         server.stop();
+         server = null;
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-web/src/main/java/org/apache/activemq/artemis/component/WebServerComponent.java:373]: Done.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java [hunk 17]
```diff
@@ -4290,6 +4295,101 @@ public boolean isStarted() {
...
+   }
+
+   @Test
+   public void testRestartEmbeddedWebServerTimeout() throws Exception {
+      FakeWebServerComponent fake = new FakeWebServerComponent(400);
+      server.addExternalComponent(fake, false);
+
+      ActiveMQServerControl serverControl = createManagementControl();
+      try {
+         serverControl.restartEmbeddedWebServer(100);
```
[the remark this answers is not in the document]
Comment (jbertram) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java]: I _think_ I understood your point. Let me know if I didn't.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/ActiveMQServerControlImpl.java [hunk 18]
```diff
@@ -4479,5 +4487,68 @@ public void replay(String startScan, String endScan, String address, String targ
...
+               startEmbeddedWebServer();
+            }
+            latch.countDown();
+         } catch (Exception e) {
+            ActiveMQServerLogger.LOGGER.failedToRestartEmbeddedWebServer(e);
+         }
+      });
+      if (!latch.await(timeout, TimeUnit.MILLISECONDS)) {
+         throw ActiveMQMessageBundle.BUNDLE.embeddedWebServerNotRestarted(timeout);
+      }
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/ActiveMQServerControlImpl.java]: Should be resolved.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/ActiveMQServerControlImpl.java [hunk 19]
```diff
@@ -4479,5 +4488,71 @@ public void replay(String startScan, String endScan, String address, String targ
...
+         } catch (Exception e) {
+            exception.set(e);
+         } finally {
+            latch.countDown();
+         }
+      });
+      boolean result = !latch.await(timeout, TimeUnit.MILLISECONDS);
+      if (result || exception.get() != null) {
+         throw ActiveMQMessageBundle.BUNDLE.embeddedWebServerNotRestarted(timeout, exception.get());
+      }
```
Comment (gemmellr) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/ActiveMQServerControlImpl.java:4534]:
```suggestion
      if (!latch.await(timeout, TimeUnit.MILLISECONDS)) {
         throw ActiveMQMessageBundle.BUNDLE.embeddedWebServerRestartTimeout(timeout);
      } else  if (exception.get() != null) {
         throw ActiveMQMessageBundle.BUNDLE.embeddedWebServerRestartFailed(exception.get());
      }
```
Could make the failure mode and exceptions clearer if they were reported separately: either the execution timed out and the state isnt actually known (it may even execute later on just fine), or it failed by throwing while we waited for execution.

(ActiveMQMessageBundle obviously would need changed to match the suggestion change)

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/ActiveMQMessageBundle.java [hunk 20]
```diff
@@ -522,4 +522,10 @@ IllegalStateException invalidRoutingTypeUpdate(String queueName,
 
    @Message(id = 229240, value = "Connection router {0} rejected the connection", format = Message.Format.MESSAGE_FORMAT)
    ActiveMQRemoteDisconnectException connectionRejected(String connectionRouter);
+
+   @Message(id = 229241, value = "Embedded web server not found")
+   ActiveMQIllegalStateException embeddedWebServerNotFound();
+
+   @Message(id = 229242, value = "Embedded web server not restarted in {0} milliseconds", format = Message.Format.MESSAGE_FORMAT)
+   ActiveMQIllegalStateException embeddedWebServerNotRestarted(long timeout, @Cause Exception e);
```
Comment (gemmellr) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/ActiveMQMessageBundle.java:530]: Using ActiveMQIllegalStateException doesnt seem quite right. Its not that the user tried to do something when it was in the wrong state, its that a result wasnt known in an allotted time, so some kind of timeout exception woudl be more appropriate.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java [hunk 21]
```diff
@@ -4290,6 +4296,106 @@ public boolean isStarted() {
...
+      server.addExternalComponent(fake, false);
+
+      ActiveMQServerControl serverControl = createManagementControl();
+      try {
+         serverControl.restartEmbeddedWebServer(10);
+         fail();
+      } catch (Exception e) {
+         Assert.assertTrue(e instanceof ActiveMQIllegalStateException);
+      }
+      startDelay.countDown();
```
Comment (gemmellr) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java:4342]: The countdown should go in a finally or it might never happen. Rather than catching Exception it could just catch the expected exception type and let everything else bubble up (then you also dont need the assertion).

File: artemis-website/package-lock.json [hunk 22]
```diff
@@ -3676,7 +3676,7 @@
       "dev": true,
       "requires": {
         "graceful-fs": "^4.1.5",
-        "minimist": "^1.2.6",
+        "minimist": "^1.2.0",
```
Comment (gemmellr) [inline artemis-website/package-lock.json:3679]: Seems like the changes in this file should go in separately if needed, they seem unrelated.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/ActiveMQServerControlImpl.java (hunk 19 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/ActiveMQServerControlImpl.java]: Yes, that would be more clear.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java (hunk 21 above)
Comment (jbertram) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java]: Done.

File: artemis-website/package-lock.json (hunk 22 above)
Comment (jbertram) [reply artemis-website/package-lock.json]: This was a mistake. Reverted.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/ActiveMQMessageBundle.java (hunk 20 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/ActiveMQMessageBundle.java]: Fair enough. Fixed.

File: artemis-website/package-lock.json [hunk 23]
```diff
@@ -3734,7 +3734,7 @@
         "boolbase": "~1.0.0",
         "css-what": "2.1",
         "domutils": "1.5.1",
-        "nth-check": "^2.0.1"
+        "nth-check": "~1.0.1"
```
Comment (gemmellr) [inline artemis-website/package-lock.json:3737]: This one seems unrelated also.

File: artemis-website/package-lock.json [hunk 24]
```diff
@@ -4899,8 +4899,7 @@
       "dev": true
     },
     "nth-check": {
-      "version": "2.0.1",
-      "resolved": "https://registry.npmjs.org/nth-check/-/nth-check-2.0.1.tgz",
+      "version": "https://registry.npmjs.org/nth-check/-/nth-check-2.0.1.tgz",
```
Comment (gemmellr) [inline artemis-website/package-lock.json:4902]: This one seems unrelated also.

**Review by gemmellr — COMMENTED** [review]
Looks good other than the couple stragglers noted.

File: artemis-website/package-lock.json (hunk 23 above)
Comment (jbertram) [reply artemis-website/package-lock.json]: Removed.
